lib/utils/promote.ts

```diff
@@ -0,0 +1,371 @@
+	}
+
+	return ip;
+}
+
+async function getOrSelectApplication(
+	sdk: resin.ResinSDK,
+	deviceType: string,
+	appName?: string,
+): Promise<resin.Application> {
+	const _ = await import('lodash');
+	const form = await import('resin-cli-form');
+	const { selectFromList } = await import('../utils/patterns');
+
+	const options: any = {
+		$expand: { user: { $select: ['username'] } },
+		$filter: { device_type: deviceType },
+	};
+
+	if (!appName) {
+		// No application specified, show a list to select one.
+		const applications = await sdk.models.application.getAll(options);
+		if (applications.length === 0) {
+			const shouldCreateApp = await form.ask({
+				message:
+					'You have no applications this device can join.\n  Would you like to create one now?',
+				type: 'confirm',
+				default: true,
+			});
+			if (shouldCreateApp) {
+				return createApplication(sdk, deviceType);
+			}
+			process.exit(1);
+		}
+		return selectFromList(
+			'Select application',
+			_.map(applications, app => _.merge({ name: app.app_name }, app)),
+		);
+	}
+
+	// We're given an application; resolve it if it's ambiguous and also validate
+	// it's of appropriate device type.
+	options.$filter = { app_name: appName };
+
+	// Check for an app of the form `user/application` and update the API query.
+	const match = appName.match(/(\w+)\/(\w+)/);
+	if (match) {
+		// These will match at most one app, so we'll return early.
+		options.$expand.user.$filter = { username: match[1] };
+		options.$filter.app_name = match[2];
+	}
+
+	// Fetch all applications with the given name that are accessible to the user
+	const applications = await sdk.pine.get<resin.Application>({
+		resource: 'application',
+		options,
+	});
+
+	if (applications.length === 0) {
+		const shouldCreateApp = await form.ask({
+			message: `No application found with name "${appName}".\n  Would you like to create it now?`,
+			type: 'confirm',
+			default: true,
+		});
+		if (shouldCreateApp) {
+			return createApplication(sdk, deviceType, options.$filter.app_name);
+		}
+		process.exit(1);
+	}
+
+	// We've found at least one app with the given name.
+	// Filter out apps for non-matching device types and see what we're left with.
+	const validApplications = applications.filter(
+		app => app.device_type === deviceType,
+	);
+
+	if (validApplications.length === 1) {
+		return validApplications[0];
+	}
+
+	// If we got more than one application with the same name it means that the
+	// user has access to a collab app with the same name as a personal app. We
+	// present a list to the user which shows the fully qualified application
+	// name (user/appname) and allows them to select
+	return selectFromList(
+		'Found multiple applications with that name; please select the one to use',
+		_.map(validApplications, app => {
+			const owner = _.get(app, 'user[0].username');
+			return _.merge({ name: `${owner}/${app.app_name}` }, app);
+		}),
+	);
+}
+
+async function createApplication(
```

Comment:
We've already got a `resin app create` command, could we use there here? Looks like the only difference is that this prompts for a name if one isn't specified, but I think there's a good argument for making `resin app create` do that too.

Reply:
It's going to be tricky to use `resin app create` directly (as in `runCapitanoCommand('app create')`) -- I need to return an `Application` instance but I don't get access to the command's output at all in order to resolve it.